Add external requests handling for ratsd core

cowbon opened the pull request:
Add the definition for endpoint `/ratsd/chares` to ratsd core, the lead attester, based on the OpenAPI YAML from Thomas F. Support for Authentication tokens and test suites will be added in later patches.

jraman567 commented on cmd/main.go:
```diff
@@ -0,0 +1,77 @@
+
+	cfg := cfg{
+		ListenAddr: DefaultListenAddr,
+		Protocol: "https",
```
Thanks for the PR, Tom. I'd like to share my thoughts on generating the API code using `oapi-codegen`.

- tools/tools.go isn't necessary since you're referring to the GitHub repo directly. So please remove the tools directory.
- I'm unsure if committing api.gen.go directly is a good idea. Since you have api/docs.go, we can run `go generate` within the api directory at some point during the build process to generate this file. That would keep the source code up to the spec in ratsd.yaml
- Since the code is auto-generated, it should be good. However, should we add functional tests for it?

Also, we need to do the following in general for the repo:

- Add a CI pipeline that runs whenever someone submits a PR
- For every commit in your PR, please run `go mod tidy` individually and ensure the mod file is up-to-date.

I still need to review the third commit. I'll wrap it up by tonight.

cowbon replied:
For https Cert and Certkey, is it fine to reuse the one from Veraison/services?

Regarding the code generation, `tools.go` is a pattern to remove `go generate` from the Makefile, see https://www.jvt.me/posts/2022/06/15/go-tools-dependency-management/. IMO the generated code is beyond the scope of the test coverage of ratsd as it should be covered by `oapi-codegen`.

thomas-fossati replied:
> For https Cert and Certkey, is it fine to reuse the one from Veraison/services?

We can reuse the infrastructure (e.g., the rootCA and scripts) but the end-entity certificate used by ratsd should be made ad-hoc, with a subject/SAN matching the expected hostname.

thomas-fossati replied:
Hi @jraman567 

> * Add a CI pipeline that runs whenever someone submits a PR

A barebones CI for the Go code has been in place since adf98a3

Until we add e2e tests, that should be enough.

Are you thinking of something different/special?

jraman567 replied:
@cowbon See https://github.com/veraison/services/blob/main/deployments/docker/src/certs/gen-certs on how Veraison generates certificates

jraman567 replied:
@thomas-fossati I think the CI for the Go code you added looks good. Thank you!

thomas-fossati replied:
On reflection, we probably need to update the go version used by the CI (currently 1.20) to reflect the requirements introduced by this PR.

jraman567 replied:
@cowbon Concerning tools.go, it's only needed if your go generate comment was like below in api/doc.go
//go:generate oapi-codegen --config=config.yaml ../docs/api/ratsd.yaml

Since you specify the full path of the oapi-codegen repo like below, the tools.go file isn't necessary to first download and build the tool
//go:generate go run github.com/oapi-codegen/oapi-codegen/v2/cmd/oapi-codegen --config=config.yaml ../docs/api/ratsd.yaml

Are you seeing any issues with removing the tools directory? I'm in favor of removing it as it doesn't appear to be doing any function presently.

jraman567 replied:
I don't know if it's best to push `api/api.gen.go` to the repo or if it's better to generate it using `go generate` during the build process.

I prefer to generate it during the build process. Something like
`go generate; go build`

@thomas-fossati Do you have any thoughts?

cowbon replied:
> On reflection, we probably need to update the go version used by the CI (currently 1.20) to reflect the requirements introduced by this PR.

Since the version bump is only introduced in this PR, I can bring it in as part of the rework

jraman567 replied:
> On reflection, we probably need to update the go version used by the CI (currently 1.20) to reflect the requirements introduced by this PR.

I agree

thomas-fossati replied:
> I don't know if it's best to push `api/api.gen.go` to the repo or if it's better to generate it using `go generate` during the build process.
> 
> I prefer to generate it during the build process. Something like `go generate; go build`
> 
> @thomas-fossati Do you have any thoughts?

The risk of regenerating on each build is that, unless we can reference a specific version of the YAML file, the business logic may break if the OpenAPI file changes.
That would add complexity each time we move the YAML, e.g., to add/tweak some functionality...

My two cents:
* Add the generated code to the repo.
* Add an explicit make rule (or script) to regenerate the code when necessary (i.e.,  when the API spec is updated).

jraman567 replied:
OK, that sounds good to me concerning code generation.

cowbon replied:
> If the default is https, do you need the default Cert & CertKey for it?

I recently added something similar to `gen_cert` in Veraison for HTTPS, but I also made the default protocol to be `http` instead of `https`. @jraman567 Does that make sense to you?

jraman567 commented on api/server.go:
```diff
@@ -0,0 +1,38 @@
+)
+
+// Ensure that we implement the server interface
+var _ ServerInterface = (*Server)(nil)
```
I don't think this line is necessary to implement the server interface. Since RatsdChares is defined below, that automatically implements the interface.

So, you can remove this line. Thanks!

jraman567 replied:
@cowbon Thanks for the PR, tom! I've completed the review and requested some changes. If you want to discuss it, I'm glad to setup a meeting. Cheers.

cowbon replied:
The type check is not deemed necessary. I removed it in the latest version of the patchset.

thomas-fossati commented on api/server.go:
```diff
@@ -0,0 +1,38 @@
+	}
+}
+
+func (s *Server) RatsdChares(w http.ResponseWriter, r *http.Request, param RatsdCharesParams) {
+	var requestData ChaResRequest
+
+	err := json.NewDecoder(r.Body).Decode(&requestData)
+	if err != nil {
+		s.logger.Error("fail to retrieve nonce from request")
+		http.Error(w, err.Error(), http.StatusBadRequest)
+		return
+	}
+
+	s.logger.Info("request nonce: ", requestData.Nonce)
+	s.logger.Info("request media type: ", *(param.Accept))
+	w.Header().Set("Content-Type", "application/eat+jwt; eat_profile=\"tag:github.com,2024:veraison/ratsd\"")
+	w.WriteHeader(http.StatusOK)
+	w.Write([]byte("hello, rastd!"))
+}
```
Question: do you want to add all the checks on input conformance (e.g., C-T, nonce format), and content negotiation in a subsequent PR along with the tests?

If so, I am OK with that, but we need to make that a bit more explicit - i.e., with one or more well-scoped tracking issue(s).

cowbon replied:
@thomas-fossati Should I go ahead to create a Github issue to state that it warrants a format check? I'd want to make clear the format of C-T and nonce to proceed on.

thomas-fossati replied:
Please, go ahead and create the issue.  The things to check are, at least:
* payload consistency (`Content-Type` header and format of the JSON payload)
* HTTP content negotiation (obey any `Accept` header)

thomas-fossati commented on ratsd.crt:
There should be a Makefile target to re-generate the chain.

cowbon replied:
`gen-certs` and `clean-certs` are now added to the Makefile.

thomas-fossati commented on cmd/config.yaml:
```diff
@@ -0,0 +1,5 @@
+logging:
+  level: debug  # valid levels: error, warning, info, debug
+ratsd:
+  listen-addr: 0.0.0.0:8888
+  protocol: http
```
Since we have the certs, I guess this can be https, no?

cowbon replied:
Yes, I've updated it to `https` in the patch that brings in the required certificates.

thomas-fossati commented on gen-certs:
```diff
@@ -0,0 +1,207 @@
+#!/bin/bash
+# Copyright 2024 Contributors to the Veraison project. SPDX-License-Identifier: Apache-2.0
```
```suggestion
# Copyright 2025 Contributors to the Veraison project.
```

cowbon replied:
Thanks for pointing this out.

thomas-fossati commented:
Thanks for the changes @cowbon!  I have added a couple of new comments inline.

iolivergithub commented:
I'm getting the following on this branch

```
ian@Debian:~/ratsd$ make
go generate ./api
go build -o ratsd ./cmd
# github.com/veraison/services/log
../go/pkg/mod/github.com/veraison/services@v0.0.2501/log/hclogger.go:119:9: cannot use &HCLogger{…} (value of type *HCLogger) as hclog.Logger value in return statement: *HCLogger does not implement hclog.Logger (missing method GetLevel)
../go/pkg/mod/github.com/veraison/services@v0.0.2501/log/hclogger.go:140:9: cannot use &HCLogger{…} (value of type *HCLogger) as hclog.Logger value in return statement: *HCLogger does not implement hclog.Logger (missing method GetLevel)
../go/pkg/mod/github.com/veraison/services@v0.0.2501/log/hclogger.go:161:9: cannot use &HCLogger{…} (value of type *HCLogger) as hclog.Logger value in return statement: *HCLogger does not implement hclog.Logger (missing method GetLevel)
make: *** [Makefile:16: build] Error 1
ian@Debian:~/ratsd$ go version
go version go1.23.2 linux/amd64
```

cowbon commented:
> I'm getting the following on this branch
> 
> ```
> ian@Debian:~/ratsd$ go version
> go version go1.23.2 linux/amd64
> an@Debian:~/ratsd$ go mod tidy
> go: downloading github.com/sagikazarmark/locafero v0.4.0
> go: downloading github.com/bytedance/sonic v1.11.3
> go: downloading github.com/goccy/go-json v0.10.2
> go: downloading github.com/json-iterator/go v1.1.12
> go: downloading go.uber.org/goleak v1.3.0
> go: downloading github.com/go-test/deep v1.0.8
> go: downloading gopkg.in/check.v1 v1.0.0-20201130134442-10cb98267c6c
> go: downloading golang.org/x/exp v0.0.0-20230905200255-921286631fa9
> go: downloading github.com/frankban/quicktest v1.14.6
> go: downloading github.com/go-playground/assert/v2 v2.2.0
> go: downloading github.com/davecgh/go-spew v1.1.2-0.20180830191138-d8f796af33cc
> go: downloading github.com/pmezard/go-difflib v1.0.1-0.20181226105442-5d4384ee4fb2
> go: downloading github.com/modern-go/reflect2 v1.0.2
> go: downloading github.com/modern-go/concurrent v0.0.0-20180306012644-bacd9c7ef1dd
> go: downloading github.com/sourcegraph/conc v0.3.0
> go: downloading github.com/kr/pretty v0.3.1
> go: downloading github.com/kr/text v0.2.0
> go: downloading github.com/rogpeppe/go-internal v1.12.0
> go: downloading github.com/chenzhuoyu/base64x v0.0.0-20230717121745-296ad89f973d
> go: downloading golang.org/x/arch v0.7.0
> go: downloading github.com/twitchyliquid64/golang-asm v0.15.1
> go: downloading github.com/chenzhuoyu/iasm v0.9.0
> go: downloading github.com/klauspost/cpuid/v2 v2.2.7
> ian@Debian:~/ratsd$ make
> go generate ./api
> no required module provides package github.com/oapi-codegen/oapi-codegen/v2/cmd/oapi-codegen; to add it:
> 	go get github.com/oapi-codegen/oapi-codegen/v2/cmd/oapi-codegen
> api/doc.go:3: running "go": exit status 1
> make: *** [Makefile:12: generate] Error 1
> ian@Debian:~/ratsd$     go get github.com/oapi-codegen/oapi-codegen/v2/cmd/oapi-codegen
> go: added github.com/dprotaso/go-yit v0.0.0-20220510233725-9ba8df137936
> go: added github.com/oapi-codegen/oapi-codegen/v2 v2.4.1
> go: added github.com/speakeasy-api/openapi-overlay v0.9.0
> go: added github.com/vmware-labs/yaml-jsonpath v0.3.2
> ian@Debian:~/ratsd$ make
> go generate ./api
> go build -o ratsd ./cmd
> # github.com/veraison/services/log
> ../go/pkg/mod/github.com/veraison/services@v0.0.2501/log/hclogger.go:119:9: cannot use &HCLogger{…} (value of type *HCLogger) as hclog.Logger value in return statement: *HCLogger does not implement hclog.Logger (missing method GetLevel)
> ../go/pkg/mod/github.com/veraison/services@v0.0.2501/log/hclogger.go:140:9: cannot use &HCLogger{…} (value of type *HCLogger) as hclog.Logger value in return statement: *HCLogger does not implement hclog.Logger (missing method GetLevel)
> ../go/pkg/mod/github.com/veraison/services@v0.0.2501/log/hclogger.go:161:9: cannot use &HCLogger{…} (value of type *HCLogger) as hclog.Logger value in return statement: *HCLogger does not implement hclog.Logger (missing method GetLevel)
> make: *** [Makefile:16: build] Error 1
> ```

Thanks Ian,
I've regenerate `go.sum` and `go.mod` and now there should be no build issue.

cowbon commented:
I accidentally closed the PR. Recreate another one.

cowbon commented:
Recreate the PR with the following fixes:
 1. HTTPS certificates 
 2. HTTP request conformance check
 3. Makefile related issues
 4. go.mod updates
